sdk/src/main/java/io/opentelemetry/sdk/trace/export/SpanWatcherProcessor.java

```diff
@@ -0,0 +1,332 @@
+/*
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.trace.export;
+
+import com.google.common.util.concurrent.MoreExecutors;
+import io.opentelemetry.internal.Utils;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanData;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * Implementation of the {@link SpanProcessor} that batches spans exported by the SDK then pushes
+ * them to the exporter pipeline.
+ *
+ * <p>All spans reported by the SDK implementation are first added to a synchronized queue (with a
+ * {@code maxQueueSize} maximum size, after the size is reached spans are dropped) and exported
+ * every {@code scheduleDelayMillis} to the exporter pipeline in batches of {@code
+ * maxExportBatchSize}.
+ *
+ * <p>If the queue gets half full a preemptive notification is sent to the worker thread that
+ * exports the spans to wake up and start a new export cycle.
+ *
+ * <p>This batch {@link SpanProcessor} can cause high contention in a very high traffic service.
+ * TODO: Add a link to the SpanProcessor that uses Disruptor as alternative with low contention.
+ */
+public final class SpanWatcherProcessor implements SpanProcessor {
```

Comment:
I think this is very similar to https://github.com/open-telemetry/opentelemetry-java/blob/master/sdk/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpansProcessor.java. Is it possible to share at least the Worker/Queuing logic?

Reply:
Yes, as stated in the description, it is probably 80% the same. But the span watcher has no queue, it has a watchlist. Spans are not removed on export. I'm not sure how easily one could share code. Also the types are slightly different since the watcher has to convert to SpanData earlier.

Reply:
As of recently, the conversion to `SpanData` is now at the same point as in the `BatchSpansProcessor`, so the types have become even more similar now.

Reply:
So if I understand correctly you will try to avoid duplicated code, is that correct?

Reply:
Hey sorry for not replying, I will leave this PR lying around some more while we are still discussing requirements for this span processor (e.g. if it should also optionally export ended spans).